Jboss cli

liquidat opened the pull request:
##### Issue Type:
- New Module Pull Request
##### Plugin Name:

web_infrastructure/jboss_cli.py
##### Summary:

This module provides JBoss management via jboss_cli for JBoss application servers.

The module on purpose uses the jboss_cli to manage the JBoss server, although this is far from perfect: the JBoss server should better be managed via API calls or similar.
However, in the field many people are used to the jabber_cli. But for such cases the generic shell or command module do not fulfill the existing needs since the average jboss_cli user often needs to set port, host, user, password and a shell environment variable. Also, these data need to be filled with sensible default values.

Please note that is is not possible to make this module check safe, since it allows for generic execution of tasks. Also , please note that changed=True is set each time a task is executed since the return value of the CLI is not specific enough.

In terms of todo, there can still a lot be improved: especially the output of data (jboss already outputs json) can be improved.
##### Example:

```
ansible host -m jboss_cli -a "command=\"/subsystem=ejb3/thread-pool=default:write-attribute(name=max-threads,value=2255)\" user=liquidat password=******
host | SUCCESS => {
    "changed": true
    "stdout": "{\"outcome\" => \"success\"}\n", 
    "stdout_lines": [
        "{\"outcome\" => \"success\"}"
    ]
}
```

mscherer commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,137 @@
+DOCUMENTATION = '''
+---
+module: jboss_cli
+version_added: "1.0"
```
This make test fail, can you switch to 2.1 ?

liquidat commented:
Yeah, I saw that in the build logs. I updated the string to 2.1.

resmo commented:
Could you please squash the commits into one?

resmo commented:
Thanks @liquidat for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

liquidat commented:
I squashed all commits, should be one single commit now.

gundalow commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,137 @@
+            jbosshome = dict(required=False, type='str', default='/usr/share/jbossas/'),
+        ),
+        # we are not there yet
+        #supports_check_mode = True,
```
Could you please add some details (in the code) here explaining why check_mode isn't available.
The comment as it stands reads like it _may_ be possible to add check_mode in the future, where as the PR description implies it isn't possible.

gundalow commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,137 @@
+    tasks:
+
+    - name: set max-threads to a given value
+      jboss_cli: command="/subsystem=ejb3/thread-pool=default:write-attribute(name=max-threads,value=3000)"
```
I believe the preferred format is `key: value`, rather than `key=value`

liquidat replied:
I took the examples from the official documentation at https://docs.jboss.org/author/display/WFLY8/CLI+Recipes - can you point me to a source about the `key: value` principle?

gundalow replied:
Hi,
You can drop this, the documentation is fine as is with `key=value`. The Ansible Core Team hasn't agreed on `key: value` yet, I was jumping the gun a little there.

gundalow commented:
Hi,
Thanks for raising a PR with this new module.

I'm not familiar with jboss, so I can't review it from that side, though the code looks sensible

Some general comments:

1) Although not a hard requirement it would be good if the code was PEP8 clean. It's a lot easier for this to be done as part of the initial module creation. If this was to be fixed after the module is submitted it makes tracing where errors got introduced a little harder.

The Makefile in ansible/ansible defines the PEP8 standard to be

`pep8 -r --ignore=E501,E221,W291,W391,E302,E251,E203,W293,E231,E303,E201,E225,E261,E241,E402 jboss_cli.py`

2) In the PR description you've implied there are a few things you'd like to fix, could you add those as details in the code so they will be available for others

Other comments are in the code diff above

liquidat commented:
Thanks, I fixed most of the points, just the `key: value` is still open.

gundalow commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,140 @@
+options:
+    user:
+        description:
+            - the name of the management user
```
Caps and fullstop on these please. I know not all modules do this.

gundalow commented:
Looks good, thanks for addressing my comments. You can ignore the `key: value` requirement.

mscherer commented:
So, the changelog is wrong, it speak of jabber_cli, not jboss_cli. That's not blocking, but i think that should be better to correct that, sorry to not have seen earlier.

liquidat commented:
@mscherer I fixed the changelog, good catch!
@bob-smith I added caps and fullstops, thanks for the hint.

gundalow commented:
Confirming all my comments have been addressed. I don't have access to JBoss, though code looks good and cleanly fails of you don't have JBoss installed.

gundalow commented:
Confirming all my comments have been addressed. I don't have access to JBoss, though code looks good and cleanly fails of you don't have JBoss installed.

shipit

bcoca commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,140 @@
+    user:
+        description:
+            - The name of the management user.
+        required: false
```
missing default, required when required is false

bcoca commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,140 @@
+DOCUMENTATION = '''
+---
+module: jboss_cli
+version_added: "2.1"
```
at this point 2.2

bcoca commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,140 @@
+        command: "deploy /path-up-to/your-app.war"
+'''
+RETURN = '''
+msg:
```
don't document default keys (msg, changed, etc), if no module specific keys just leave comment 

`# only returns defaults`

bcoca commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,140 @@
+            - The JBoss path environment variable.
+        required: false
+        default: '/usr/share/jbossas/'
+    command:
```
this 'command' facility is something we are trying to avoid, it is not imperative and really does not seem to add much value, I would prefer to see specific options or even different modules to handle deployment or attribute updates.

liquidat commented:
@bcoca I have fixed the version number, added the missing defaults and streamlined the examples so that they do not show default values anymore.
Regarding the general issue that this is a command module: JBoss ships with a command, and I want to make it easier to automate it. I will try to focus on an API module as well, but from my point of view these are two different things. However, I would understand if that this module does not fit into Ansible's strategy right now, just let me know.

gregdek commented:
Thanks again to @liquidat for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks again to @liquidat for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@liquidat A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@liquidat Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

liquidat commented:
@gregdek I think all pending actions have been resolved with my update from May 5th. Can you please re-check if any actions are missing and if, point out which?

gundalow commented:
@liquidat As statedin Gregdek's comment you need to add a comment of "ready_for_review" to get this back into review.

gundalow commented:
Hi,

Looks like there is a small tweak needed to support Python 3

```
python3.5 -m compileall -fq jboss_cli.py 
*** Error compiling 'jboss_cli.py'...
  File "jboss_cli.py", line 127
    except subprocess.CalledProcessError, cpe:
                                        ^
SyntaxError: invalid syntax

```

The following guide should explain what's needed http://docs.ansible.com/ansible/developing_modules_python3.html

Thanks

liquidat commented:
@gundalow Thanks for the feedback. I fixed the exception handling to be Python 3 compatible (had the same problem already in another module).

liquidat commented:
ready_for_review

gundalow commented:
Thanks for the quick turn around

Confirming all my comments have been addressed. I don't have access to JBoss, though code looks good and cleanly fails of you don't have JBoss installed.

gundalow commented:
shipit

gundalow commented:
Looks like Git has gotten a bit confused by this. Can you please rebase this PR that should make Travis green, after that we should be able to get it merged

liquidat commented:
@gundalow Well, that rebase certainly didn't work as expected. I think I screwed up the Git history pretty much of this PR. I'll try to fix this - though git rebase is driving me nuts right now...
Besides, I am surprised that the travis checks fail again - at a yum plugin?!

gregdek commented:
Thanks again to @liquidat for this PR, and thanks @banyek, @lpenz, @ritzk, @chrrrles, @brian-brazil, @dstoflet, @srvg, @UnderGreen, @andsens, @sayap, @jpmens, @eikef, @ansible, @Java1Guy, @fabulops, @robinro, @tarka, @TimothyVandenbrande, @ramondelafuente, @alcamie101, @ccollicutt, @quentinsf, @cloudnull, @rosmo, @bpennypacker, @ignatenkobrain, @DJMuggs, @privateip, @mcltn, @matze, @indrajitr, @emonty, @nibalizer, @rickmendes, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, @pb8226, @jirutka, @AugustusKling, @whenrik, @michaeljs1990, @wimnat, @makaimc, @jhawkesworth, @dschep, @mattdavispdx, @resmo, @dazworrall, @andreparames, @manuel-sousa, @chrishoffman, @romanek-adam, @dohoangkhiemgmail.com, @thoiberg, @jtyr, @joelthompson, @jpmens, @sivel, @NickatEpic, @joshludwig, @smadam813, @drybjed, @JoergFiedler, @Etherdaemon, @willybarro, @ravibhure, @schwartzmx, @willthames, @wimnat, @DavidWittman, @mulby, @skornehl, @mcodd, @bcoca, @dermute, @pcgentry, @olsaki, @dagwieers, @jhoekx, @abulimov, @erydo, @treyperry, @ovcharenko, @pyykkis, @trondhindenes, @mhite, @caphrim007, @mjschultz, @mgruener, @petemounce, @loia, @dmtrs, @resmo, @mcv21, @jcpowermac, @mtnbikenc, @stygstra, @slok, @jirutka, @luisperlaz, @sgargan, @maxamillion, @hkariti, @LinusU, @jsmartin, @drewkerrigan, @ThePixelDeveloper, @chrisisbeef, @tanpeter, @Etienne-Carriere, @dj-wasabi, @lujeni, @mscherer, @tksmd, @drcapulet, @weaselkeeper, @minichate, @dougluce, @joshainglis, @karmab, @dankeder, @amir343, @n0ts, @franckcuny, @groks, @L2G, @szinck, @troy2914 for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

liquidat commented:
Fixed the rebase as requested. Thanks to @bcoca for the help.

gregdek commented:
Thanks @liquidat for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gundalow commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,133 @@
+
+    try:
+        rc, stdout, stderr = module.run_command(cmd)
+        # TODO: Prettify output; currently JSON similar, but not JSON compatible
```
How much would it take to do this?

liquidat replied:
@gundalow That is really hard to say. So far I was not able to find a way to make JBoss change its output. Also, there is no readily available script to integrate I know of. It _might_ be that a simple substitution could work, but I am not sure if there are hidden caveats in the JBoss format.

gundalow commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,133 @@
+        rc, stdout, stderr = module.run_command(cmd)
+        # TODO: Prettify output; currently JSON similar, but not JSON compatible
+        module.exit_json(changed=True, stdout=stdout)
+    except subprocess.CalledProcessError as cpe:
```
Travis is showing a Python syntax error for this against Python 2. The module needs to be compatible with Python 2 & 3.

If you think the Python 3 porting guide isn't clear feel free to raise a PR to update the wording or let me know how we can make it clearer and I'll update it. 

Thanks

sm4rk0 replied:
The document says

> we have to use

so, I suppose we have no other choice if we want it to compile with 2.4

gundalow commented:
Git can sometimes be a pain, it is possible somethings to get into strange situations. Thank you for fixing that. 

needs_revision

gundalow commented:
@liquidat For future reference this guide shows how to update your branch cleanly. I should have mentioned this in my previous message.

http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

liquidat commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,133 @@
+        rc, stdout, stderr = module.run_command(cmd)
+        # TODO: Prettify output; currently JSON similar, but not JSON compatible
+        module.exit_json(changed=True, stdout=stdout)
+    except subprocess.CalledProcessError as cpe:
```
@gundalow This syntax fails with Python 2.4, but not with Python 2.6 - I can change the exception handling like https://docs.ansible.com/ansible/developing_modules_python3.html suggests. But as the guide states, that is an "ugly construction" ;-)
So please point out if you think it is worth it.

liquidat commented:
@gundalow Thanks for the link!

gregdek commented:
Ping @liquidat -- still waiting on rebase, thanks!

alikins commented on web_infrastructure/jboss_cli.py:
```diff
@@ -0,0 +1,133 @@
+        # TODO: Prettify output; currently JSON similar, but not JSON compatible
+        module.exit_json(changed=True, stdout=stdout)
+    except subprocess.CalledProcessError as cpe:
+        module.fail_json(msg=str(dir(cpe)))
```
Not sure I understand this construct.

Why not just returnstr(cpe) for msg?  Or build a message with cpe's attributes?

liquidat replied:
This was copied from the existing `system/alternatives.py` which I used as a template. Not sure why it was done exactly that way, but I figured it would be better to copy what is already there than introducing new ways.

gregdek commented:
@liquidat A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

liquidat commented:
I've fixed the mentioned Python 3 exception handling. Should be fine now.
ready_for_review

mscherer commented:
Now, it fial on python 2.4. See https://docs.ansible.com/ansible/developing_modules_python3.html for the right idiom for exception.
